src/components/Firestore/Document.tsx

```diff
@@ -19,28 +19,47 @@ import './index.scss';
 import { firestore } from 'firebase';
 import { Typography } from '@rmwc/typography';
 import { Icon } from '@rmwc/icon';
+import { Route, useRouteMatch } from 'react-router-dom';
 
 import CollectionList from './CollectionList';
 import DocumentPreview from './DocumentPreview';
+import Collection from './Collection';
+import { useApi } from './ApiContext';
 
 export interface Props {
-  reference: firestore.DocumentReference;
+  reference?: firestore.DocumentReference;
```

Comment:
This feels hacky to me and introduces a whole bunch of conditionals in the render body. For example, the type system doesn't reflect the fact that we'll never have `snapshot` but not `reference`, and if we ever have a null error it can be hard to debug since it now shows as root.

Let's actually make two different components, one for `<DatabaseRoot>` and another for `<Document>` and extract the common things include routes into a lower-level component. This will also greatly simply future work with missing documents.

Reply:
I could type this to be more narrow... no-reference/no-snapshot, reference,maybe-snapshot

Or do you think HOCs are the way to go?

Reply:
I think we can get this done by extracting helper components that are present in both root and document. That component can probably leave `children` open for data viewer etc. I think that will be cleaner and less mental overhead than messing with typing. I don't see an immediate need for HOC, but feel free to prove me wrong.

Reply:
Ack... I updated Document.tsx, take a look. There is now Root/Document components that use an underlying private Doc.